---
title: 'Plot comparing differential expression of LF82wt vs LF82 mutant and MG1655 mutant vs MG1655 wt'
author: "Hernan Lorenzi"
date: "02/06/2024"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

#### Load libraries

```{r load libraries}
pacman::p_load(AnnotationDbi,pheatmap,EnhancedVolcano,ggpubr,DESeq2,stringr,biomaRt,tidyverse,pcaExplorer,VennDiagram,clusterProfiler,GOSemSim,ggsci,viridis,ggrepel,RColorBrewer,msigdbr,cowplot,enrichplot,ggupset,broom,ggraph)

```


```{r}
#
# --- function for PCA plots ---
#
plot_PCA = function(object, color_by="condition", 
                    shape_by = 19, ntop=500, size = 3,
                    returnData=FALSE, pcs = c(1,2))
{
  # Check variables are present in object
  intgroup = c(color_by)
  if (shape_by != 19){intgroup <- c(intgroup, shape_by)}
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )

  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  d <- data.frame(PC1=pca$x[,pcs[1]], PC2=pca$x[,pcs[2]], group=group, intgroup.df, name=colnames(object))
  colnames(d)[1] <- paste0("PC",pcs[1])
  colnames(d)[2] <- paste0("PC",pcs[2])
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:2]
    return(d)
  }
  
  ggplot(data=d, aes_string(x=colnames(d)[1], y=colnames(d)[2], color=color_by, shape=shape_by)) + 
    geom_point(size=size) + 
    scale_color_lancet() +
    xlab(paste0("PC",pcs[1],": ",round(percentVar[pcs[1]] * 100, digits = 2),"% variance")) + # fixed
    ylab(paste0("PC",pcs[2],": ",round(percentVar[pcs[2]] * 100, digits = 2),"% variance")) + # fixed
    coord_fixed(ratio = (max(d[,1])-min(d[,1]))/(max(d[,2])-min(d[,2])))
}

```



```{r}
log2fc_lf82_mg <- read.table(file = "./data/Log2FC_LF82_MG.txt", header = T, sep = "\t", row.names = 1)

genes_of_interest <- c("metE", "yobH", "lptM", "yigA", "xerC", "yigB", "ampC", "yciY", "shiA", "acrZ", "sltY", "fucI", "ybdG", "mlaA", "cysZ", "dpaA", "xseA", "mepK", "aroP", "mnmC", "rsuA", "bcr", "csrD", "sppA", "gltP", "ruvA", "yacC", "folD")

# Generate first plot with all genes
my_title <- "Log2(Fold Change) of differential gene expression\nLF82-WT/LF82-Mut and MG1655-Mut/MG1655-WT"

df <- log2fc_lf82_mg

# Invert Log2FC for LF82 to reflect WT/Mut rather than the original data Mut/WT
df$Log2FC_LF82 <- df$Log2FC_LF82 * -1

# remove rows where Log2FC LF82/MG == NA
df <- df[!c(is.na(df$Log2FC_MG) | is.na(df$Log2FC_LF82)), ]

df$padj_LF82[is.na(df$padj_LF82)] <- 1
df$padj_MG[is.na(df$padj_MG)] <- 1
df$min_pval <- apply(dplyr::select(df,c("padj_LF82","padj_MG")), 1, function(x){ min(x)})
df <- df[order(df$min_pval, decreasing = T),]
df$significance <- as.factor(apply(dplyr::select(df,c("padj_LF82","padj_MG")), 1, function(x){ ifelse( (x[1] <= 0.05 & x[2] <= 0.05), "Sig both", ifelse( (x[1] <= 0.05), "Sig LF82", ifelse( (x[2] <= 0.05), "Sig MG1655","N.S."))) }))

myColor <- c("N.S."="gray","Sig both"="orange3","Sig LF82"="green4","Sig MG1655"="yellow3")

p4 <- ggplot(df, aes(x=Log2FC_LF82, y=Log2FC_MG, label = gene_name_MG)) + 
  labs(title = my_title) +
  xlab(bquote('LF82 '* ~Log[2]*'(Fold Change)')) +
  ylab(bquote('MG1655 '* ~Log[2]*'(Fold Change)')) +
  geom_point(aes(colour=significance), size = 1, alpha = 1 ) + 
  ylim(-3.5,3.5) + #ylim(min(df$Log2FC_MG),max(df$Log2FC_MG)) + 
  xlim(-3.5,3.5) + #xlim(min(df$Log2FC_LF82),max(df$Log2FC_LF82)) +
  geom_abline(slope = 1, intercept = 0, col = "black", size=0.5, linetype="dashed") +
  scale_x_continuous(breaks =c(-3,-2,-1,0,1,2,3)) +
  scale_y_continuous(breaks =c(-3,-2,-1,0,1,2,3)) +
  theme_minimal() + 
  scale_colour_manual(values = myColor) +
  theme(
    legend.direction = "vertical",  
    legend.position = c(.25, .95),
    legend.justification = c("right", "top"),
    legend.box.just = "right",
    legend.margin = margin(3, 3, 3, 3),
    legend.text = element_text(size = 8),
    legend.title = element_text(face = "bold" ,size = 8, vjust = 0.9))

p4 <-  p4 + guides(color = guide_legend(override.aes = list(size = 5)))

# Change legend title
p4$labels$colour = paste("Significance" )

#genes_of_interest
my_list <- genes_of_interest # genes_of_interest.names$gene_name

df.subset <- subset(df, gene_name_LF82 %in% my_list)


p4 <- p4 + geom_point(data = df.subset, 
                      aes(x = Log2FC_LF82, y = Log2FC_MG, fill=significance), 
                      size = 3, 
                      pch=21, 
                      colour="black", 
                      show.legend = FALSE) + 
            scale_fill_manual(values = myColor) 

#ggsave2(filename = "Log2FC_dotplot_from_orig_DE_no_labels.pdf", plot = p4, path = "./Plots", width = 10, height = 10)

# Adding labels to genes of interest
p4 <- p4 + geom_text_repel(data = df.subset, 
                            aes(label = gene_name_MG),
                            colour = "black",
                            box.padding   = 0.4, 
                            point.padding = 0.3,
                            segment.color = 'black',
                            na.rm = TRUE,
                            size = 5,
                            min.segment.length = 0.02,
                            direction = "both",
                            segment.curvature = -0.1,
                            segment.ncp = 3,
                            segment.angle = 20,
                            max.iter = 1e6, 
                            max.overlaps = 20,
                           max.time = 10)

#ggsave2(filename = "Log2FC_dotplot_from_orig_DE.pdf", plot = p4, path = "./Plots", width = 10, height = 10)
p4

```


```{r}
print(sessionInfo())
```
