---
title: ' RNAseq Analysis E. coli strain MG1655 RNAseq on MG1655 reference and LF82 SD105 on LF82 reference'
author: "Hernan Lorenzi"
date: "12/06/2023"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

#### Load libraries

```{r load libraries}
pacman::p_load(AnnotationDbi,pheatmap,EnhancedVolcano,ggpubr,DESeq2,stringr,biomaRt,tidyverse,pcaExplorer,VennDiagram,clusterProfiler,GOSemSim,ggsci,viridis,ggrepel,RColorBrewer,msigdbr,cowplot,enrichplot,ggupset,broom,ggraph)

```

#### Define functions

```{r aux functions}

# Load auxyliary functions
source(file = "../results_MG1655/01_aux_rnaseq_functions.R")

# Load enrichment functions
source(file = "../results_MG1655/02_Gene_enrichment_functions.R")

```

```{r}
#
# --- function for PCA plots ---
#
plot_PCA = function(object, color_by="condition", 
                    shape_by = 19, ntop=500, size = 3,
                    returnData=FALSE, pcs = c(1,2))
{
  # Check variables are present in object
  intgroup = c(color_by)
  if (shape_by != 19){intgroup <- c(intgroup, shape_by)}
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )

  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
  
  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  d <- data.frame(PC1=pca$x[,pcs[1]], PC2=pca$x[,pcs[2]], group=group, intgroup.df, name=colnames(object))
  colnames(d)[1] <- paste0("PC",pcs[1])
  colnames(d)[2] <- paste0("PC",pcs[2])
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:2]
    return(d)
  }
  
  ggplot(data=d, aes_string(x=colnames(d)[1], y=colnames(d)[2], color=color_by, shape=shape_by)) + 
    geom_point(size=size) + 
    scale_color_lancet() +
    xlab(paste0("PC",pcs[1],": ",round(percentVar[pcs[1]] * 100, digits = 2),"% variance")) + # fixed
    ylab(paste0("PC",pcs[2],": ",round(percentVar[pcs[2]] * 100, digits = 2),"% variance")) + # fixed
    coord_fixed(ratio = (max(d[,1])-min(d[,1]))/(max(d[,2])-min(d[,2])))
}

```

#### Load read counts data MG1655

```{r Loading data}
all.star <- read.delim2("./data/read_counts_mg_lf82.txt", sep = "\t", header = TRUE, row.names = 1, comment.char = c("#") )

format_star <- function(star_file){
  names(star_file) <- names(star_file) %>% 
    str_remove_all(pattern = "results.03map_reads.|.Aligned.sortedByCoord.out.bam") 
  return(star_file[3:ncol(star_file)])
}

# Format star counts file
all <- format_star(star_file = all.star)

# Make sure read counts are numeric and rounded to 0 decimals
all.tmp <- as.data.frame(lapply(all, function(x){ round(as.numeric(x), digits = 0)} ))
rownames(all.tmp) <- rownames(all)
all <- all.tmp

#Remove all zero rows
#all <- remove_all_zero_rows(all, min_total_count = 0)

column_names <- names(all) %>% sort()

all <- select(all, column_names)
```

#### Make metadata table from 'all'

```{r}
# Load metadata

metadata <- data.frame(row.names = column_names, 
                       genotype = c(rep("WT", 3),rep("Mutant",3),rep("Mutant", 3),rep("WT",3)), 
                       strain = c(rep("LF82",6),rep("MG1655",6)))


# Add total read counts and sample id columns to metadata
metadata$read_counts <- colSums(all)

# Add "Sample_name" as column in metadata
metadata$sample_name <- rownames(metadata)

# Add column combining genotype and treatment
metadata$group <- paste(metadata$strain,metadata$genotype, sep = "_")

```


#### Normalize data to TPMs to run some comparative analysis across samples

```{r}
# Using annotation version GRCm39 (current)
#all.tpm <- normalize_by_TPM(counts.df = all, 
#                            gene_length = dplyr::select(all.star, c("Length")))
```

########################################### 

# Analysis of expression data using DESeq2

########################################### 

```{r deseq2}
# Convert metadata to factors
for (variable in c("genotype", "strain","sample_name","group")){
  metadata[,variable] <- as.factor(str_replace_all(metadata[,variable], pattern = " ", replacement = "_"))  
}

```

# Analysis of Dataset

```{r Dataset one}
# Generate DESeq2 object for NS and ST condition ONLY. We could potentially add Read_counts as either a covariate or as a factor (by grouping Read counts into Low and High count bins). Ideally, this should be done adding Read_counts as a random effect using mixed models, that can only be applied in Limma package. Adding it as a fixed effect in the design formula might lower the power of the analysis.

dir.create(path = "./Plots", showWarnings = FALSE)

# Create DESeq object
dds.all <- DESeqDataSetFromMatrix(countData = all, 
                              colData = metadata,  
                              design = ~ strain + genotype)

# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "sample_name", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "group", 
          title = "Total read counts", 
          ylab = "Read count",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc") 
ggsave2("Plots/barplot_read_counts.pdf", plot = p)
print(p)
```

```{r}
# Normalize counts
vsd.one <- vst(dds.all, blind=FALSE)
rlog.one <- rlog(dds.all, blind=FALSE)

# Keep genes with at least 20 reads total across samples
#keep <- rowSums(counts(dds.all)) >= 20
#dds.all <- dds.all[keep,]

# Calculate distances between samples
sampleDists <- dist(t(assay(vsd.one)))

# Plot inter-sample distances
old.par <- par(no.readonly=T)

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rlog.one$strain, rlog.one$genotype, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
p.pheatmap <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

ggsave2(filename = "unsupervised_clustering_rnaseq_profile_20plus_reads.pdf", plot = p.pheatmap, path = "./Plots")
print(p.pheatmap)
```

```{r}
dds_res <- list()

dds_res <- dds.all #[ , dds.all$Tissue=="all_data"]

rlog_res <- list()
rlog_res <- rlog(dds_res, blind=FALSE)

# Filter out genes that are expressed in LF82/MG1655 strain but not the other
keep <- apply(assay(rlog_res), 1, function(x){sum(x>0) > 6})
rlog_res.all_strains <- rlog_res[keep,]

# PCA
rlog.one <- rlog_res

# PC1 - PC2
principal_components <- c(1,2)
pca_12.p <- plot_PCA(object = rlog.one,color_by = "genotype", shape_by = "strain", returnData = FALSE ,pcs = principal_components)

# PC2 - PC3
principal_components <- c(2,3)
pca_23.p <- plot_PCA(object = rlog.one,color_by = "genotype", shape_by = "strain", returnData = FALSE ,pcs = principal_components)

# PC1 - PC3
principal_components <- c(1,3)
pca_13.p <- plot_PCA(object = rlog.one,color_by = "genotype", shape_by = "strain", returnData = FALSE ,pcs = principal_components)

ggsave(paste0("Plots/pca_PC12_Group.pdf"), plot = pca_12.p)
print(pca_12.p)
ggsave(paste0("Plots/pca_PC13_Group.pdf"), plot = pca_13.p)
print(pca_13.p)
ggsave(paste0("Plots/pca_PC23_Group.pdf"), plot = pca_23.p)
print(pca_23.p)

```

PCA analysis shows that samples separate by genotype and treatment.

### resultsNames(dds)

#### Filtering out poorly-expressed genes (less than 20 reads across all samples)

```{r filtering genes based on total counts of reads}

# Keep genes with at least 10 reads total across samples
keep <- rowSums(counts(dds_res)) >= 20
dds_res <- dds_res[keep,]
all <- all[keep,]
all.star <- all.star[keep,]
```

##### Using groups instead of interactions

```{r}
ensembl_to_symbol <- read.delim(file = "./data/gene_names mg_lf82.txt", col.names = c("Ensembl_ID","gene_name"), header = F)

# Save sorted files as a list
DE_results <- list()
geneids.DE <- list()

# Define function for processing and saving result tables
sort_and_write_res_table <- function(result_table, file_name){
  dir.create(path = "./DE", showWarnings = FALSE)
  # Sort genes by (padj)
  result_table_sorted <- result_table[order(result_table$padj, decreasing = FALSE),]
  # Add gene symbols
  gene_list <- rownames(result_table_sorted)
  symbol_list <- ensembl_to_symbol$gene_name[match(gene_list, ensembl_to_symbol$Ensembl_ID)]
  df <-as.data.frame(cbind(result_table_sorted, Gene_name = symbol_list))
  
  # Write sorted table to file
  write.table(df, file = paste0("./DE/",file_name,".txt"), 
            sep = "\t", col.names=NA)
  return(df)
}
```

```{r}
# Calculate DE for all_data samples
#design(dds.rnaseA) <- ~Treatment # Removid Read.depth from formula given that all samples are Read.depth = High

design(dds_res) <- ~group
dds_res$group <- relevel(dds_res$group, "LF82_Mutant")
dds_res <- DESeq(dds_res)
my_contrasts <- resultsNames(dds_res)
res_WT_vs_Mut_LF82 <- lfcShrink(dds_res, contrast=c("group", "LF82_WT", "LF82_Mutant"), type = "ashr", )
res_WT_vs_Mut_LF82$LF82_IDs <- all.star$LF82_IDs

dds_res$group <- relevel(dds_res$group, "MG1655_WT")
dds_res <- DESeq(dds_res)
my_contrasts <- resultsNames(dds_res)
res_Mut_vs_WT_MG1655 <- lfcShrink(dds_res, contrast=c("group", "MG1655_Mutant", "MG1655_WT"), type = "ashr", )
res_Mut_vs_WT_MG1655$MG1655 <- all.star$MG1655_IDs

summary(res_WT_vs_Mut_LF82, alpha = 0.05)
summary(res_Mut_vs_WT_MG1655, alpha = 0.05)

# Sort results by Log2FC
res_WT_vs_Mut_LF82_sorted <- sort_and_write_res_table(result_table = res_WT_vs_Mut_LF82, file_name = paste0("res_WT_vs_Mut_LF82")) 
res_Mut_vs_WT_MG1655_sorted <- sort_and_write_res_table(result_table = res_Mut_vs_WT_MG1655, file_name = paste0("res_Mut_vs_WT_MG1655")) 



```

```{r}
table_counts_normalized <- counts(dds_res, normalized=TRUE)
write.table(x = as.data.frame(table_counts_normalized), file = "read_counts_deseq2_normalized.txt", sep = "\t", col.names = NA)
```

#### Genes of interest
```{r}
# yifL = lptM
# ybgD b0719 has less than 20 reads across all samples
genes_of_interest <- c("metE","ampC","fucI","aceB","shiA", "ybgD", "mlaA", "cysZ", "acrZ", "bcr","sppA","yifL","dapF","yigA","xerC","yigB","yobH","ruvA","zapC","mdtK")
```



#### Generate volcano plots
```{r}

volcano_plot_with_ids <- function(res.tmp, log_scale = FALSE, gene_list){
  vp <- EnhancedVolcano(res.tmp, 
                        lab = res.tmp$Gene_name, 
                        x = 'log2FoldChange', 
                        y = 'padj',
                        pCutoff = 0.05,
                        FCcutoff = 1,
                        pointSize = 1,
                        colAlpha = 4/5,
                        labSize = 3,  # Controls labels size
                        labCol = "black",
                        title = '',
                        titleLabSize = 10,
                        subtitle = '', # add subtitle here
                        subtitleLabSize = 10,
                        legendPosition = 'right',
                        legendLabSize = 10,
                        legendIconSize = 4.0,
                        axisLabSize = 10,
                        drawConnectors = TRUE,
                        selectLab = gene_list, # vector of gene symbols to label on volcanoplot
                        boxedLabels = FALSE,
                        gridlines.major = FALSE,
                        gridlines.minor = FALSE,
                        hlineCol = "gray", vlineCol = "gray"
  )
  
  #theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
  
  if (log_scale){
    vp <- vp + scale_x_log10()
  }
  
  
  return(vp)
}


vp1 <- volcano_plot_with_ids(res.tmp = res_WT_vs_Mut_LF82_sorted,
                               log_scale = FALSE, 
                               gene_list = genes_of_interest)
                               
ggsave(filename = paste0("./Plots/WT_vs_Mut_LF82_VolcanoPlot.pdf"), 
       plot = vp1, width = 6, height = 6)

vp2 <- volcano_plot_with_ids(res.tmp = res_Mut_vs_WT_MG1655_sorted, 
                               log_scale = FALSE, 
                               gene_list =genes_of_interest
                               )

ggsave(filename = paste0("./Plots/Mut_vs_WT_MG1655_VolcanoPlot.pdf"), 
       plot = vp2, width = 6, height = 6)

print(vp1)
print(vp2)
```

#### Plot heatmaps
```{r}
#genes_of_interest.ensmbl <- rownames(head(assay(rlog.one)))
#ensembl_to_symbol.bkp <- ensembl_to_symbol
ensembl_to_symbol <- subset(ensembl_to_symbol, Ensembl_ID %in% rownames(assay(rlog.one)))

genes_of_interest.names<- subset(ensembl_to_symbol, gene_name %in% genes_of_interest)


# Specify colors
ann_colors = list(
    genotype = c(Mutant = "white", WT = "firebrick"),
    strain = c(LF82 = "#1B9E77", MG1655 = "#D95F02"))

annot_col <- as.data.frame(dplyr::select(metadata,c("genotype","strain")))

p1.heatmap <- pheatmap(assay(rlog.one)[genes_of_interest.names$Ensembl_ID,], 
          cluster_rows=T, 
          show_rownames=TRUE,
          cluster_cols=T, 
          annotation_col = annot_col, 
         labels_row = genes_of_interest.names$gene_name, 
         labels_col = metadata$group, 
         annotation_colors = ann_colors)

p2.heatmap <- pheatmap(assay(rlog.one)[genes_of_interest.names$Ensembl_ID,], 
          cluster_rows=T, 
          show_rownames=TRUE,
          cluster_cols=T, 
          annotation_col = annot_col,
          annotation_colors = ann_colors)

ggsave2(filename = "mut_vs_wt_heatmap.pdf", plot = p1.heatmap, path = "./Plots", width = 7, height = 5)
```
```{r}
res_WT_vs_Mut_LF82.sig <- subset(res_WT_vs_Mut_LF82, padj <= 0.05 & abs(log2FoldChange) >= 1)
res_Mut_vs_WT_MG1655.sig <- subset(res_Mut_vs_WT_MG1655, padj <= 0.05 & abs(log2FoldChange) >= 1) 
sign_genes <- unique(c(rownames(res_WT_vs_Mut_LF82.sig),rownames(res_Mut_vs_WT_MG1655.sig)))
rlog_res.all_strains.names <- rownames(rlog_res.all_strains@assays@data@listData[[1]])
sign_genes <- sign_genes[sign_genes %in% rlog_res.all_strains.names]

p3.heatmap <- pheatmap(assay(rlog_res.all_strains)[sign_genes,], 
          cluster_rows=T, 
          show_rownames=TRUE,
          cluster_cols=T, 
          annotation_col = annot_col,
          annotation_colors = ann_colors, fontsize_row = 8)

print(p3.heatmap)

ggsave2(filename = "mut_vs_wt_heatmap_sign.pdf", plot = p3.heatmap, path = "./Plots", width = 7, height = 20)

sign_genes.both_comp <- c(rownames(res_WT_vs_Mut_LF82.sig),rownames(res_Mut_vs_WT_MG1655.sig))[duplicated(c(rownames(res_WT_vs_Mut_LF82.sig),rownames(res_Mut_vs_WT_MG1655.sig)))]

sign_genes.both_comp <- sign_genes.both_comp[sign_genes.both_comp %in% rlog_res.all_strains.names]

rlog_res.all_strains.both <- rlog_res.all_strains[sign_genes.both_comp,]

# gene names
ensembl_to_symbol.both_comp <- subset(ensembl_to_symbol, Ensembl_ID %in% sign_genes.both_comp)
rownames(ensembl_to_symbol.both_comp) <- ensembl_to_symbol.both_comp$Ensembl_ID

ensembl_to_symbol.both_comp <- select(ensembl_to_symbol.both_comp, "gene_name")

sort.idx <- match(rownames(ensembl_to_symbol.both_comp),
                        rownames(assay(rlog_res.all_strains.both)) 
                        )

genes_names <- ensembl_to_symbol.both_comp$gene_name[order(sort.idx)]

p4.heatmap <- pheatmap(assay(rlog_res.all_strains.both), 
          cluster_rows=T, 
          show_rownames=TRUE,
          cluster_cols=T, 
          labels_row = genes_names,
          annotation_col = annot_col,
          annotation_colors = ann_colors,
          fontsize_row = 8)

print(p4.heatmap)

ggsave2(filename = "mut_vs_wt_heatmap_both_sign.pdf", plot = p4.heatmap, path = "./Plots", width = 7, height = 10)
```


```{r}
print(sessionInfo())
```
